### PR TITLE
mkShell: make it buildable

### PR DESCRIPTION
**Pull request by zimbatm**
When I designed `mkShell`, I didn't have a good idea of what the output
should look like and so decided to make the build fail. In practice,
this causes quite a bit of confusion and complications because now the
shell cannot be part of a normal package set without failing the CI as
well.

This commit changes that build phase to record all the build inputs in a
file. That way it becomes possible to build it, makes sure that all the
build inputs get built as well, and also can be used as a GC root.
(by applying the same trick as #95536).

The documentation has also been improved to better describe what mkShell
does and how to use it.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
You could also write out a shell script that behaves like a better `nix-shell`.

Right now, the output is a single file with no room for extension, because it is not a directory. A good first step is to write to `$out/env.sh` instead, so other files can be added.

**Comment by zimbatm** [comment]
There is an infinite number of things I could do. The question is; is this PR better than before? I will make it so that the file is unusable, as it's not meant to be part of the interface.

**Comment by roberth** [comment]
I'm not asking you to do an infinite number of things, just to prevent a future breaking change.

> I will make it so that the file is unusable

That makes it a non-breaking change. :+1:

File: doc/builders/special/mkshell.section.md [hunk 1]
```diff
@@ -1,17 +1,37 @@
...
+    export DEBUG=1
+  '';
 }
 ```
+
+## Attributes
+
+* `name` (default: `nix-shell`). Set the name of the derivation.
+* `packages` (default: `[]`). Add build dependencies to the `nix-shell` environment.
+* `inputsFrom` (default: `[]`). Add build dependencies of the dependencies to the `nix-shell` environment.
```
Comment (roberth) [inline doc/builders/special/mkshell.section.md:27]:
```suggestion
* `inputsFrom` (default: `[]`). Add build dependencies of the listed derivations to the `nix-shell` environment.
```

File: doc/builders/special/mkshell.section.md [hunk 2]
```diff
@@ -1,17 +1,37 @@
...
+  shellHook = ''
+    export DEBUG=1
+  '';
 }
 ```
+
+## Attributes
+
+* `name` (default: `nix-shell`). Set the name of the derivation.
+* `packages` (default: `[]`). Add build dependencies to the `nix-shell` environment.
```
Comment (roberth) [inline doc/builders/special/mkshell.section.md:26]:
```suggestion
* `packages` (default: `[]`). Add executable packages to the `nix-shell` environment.
```

Dependencies can be `buildInputs`, but these are `nativeBuildInputs`, which are executable.

File: doc/builders/special/mkshell.section.md [hunk 3]
```diff
@@ -1,17 +1,37 @@
...
+* `inputsFrom` (default: `[]`). Add build dependencies of the dependencies to the `nix-shell` environment.
+* `shellHook` (default: `""`). Bash script that is executed by `nix-shell`.
```
Comment (roberth) [inline doc/builders/special/mkshell.section.md:28]:
```suggestion
* `shellHook` (default: `""`). Bash statements that are executed by `nix-shell`.
```

File: pkgs/build-support/mkshell/default.nix [hunk 4]
```diff
@@ -41,10 +41,14 @@ stdenv.mkDerivation ({
...
-  nobuildPhase = ''
-    echo
-    echo "This derivation is not meant to be built, aborting";
-    echo
-    exit 1
+  phases = [ "buildPhase" ];
+
+  buildPhase = ''
+    echo "-------------------------------------------" >> $out
+    echo " Do not use this file. See the mkShell doc." >>$out
```
Comment (roberth) [inline pkgs/build-support/mkshell/default.nix:48]: I don't quite understand why.

Comment (zimbatm) [reply pkgs/build-support/mkshell/default.nix]: Ok, this is ready for another round of review.

File: pkgs/build-support/mkshell/default.nix [hunk 5]
```diff
@@ -41,10 +41,15 @@ stdenv.mkDerivation ({
...
-  nobuildPhase = ''
-    echo
-    echo "This derivation is not meant to be built, aborting";
-    echo
-    exit 1
+  phases = [ "buildPhase" ];
+
+  buildPhase = ''
+    echo "------------------------------------------------------------" >>$out
+    echo " WARNING: this file existence or location is not guaranteed." >>$out
```
Comment (cole-h) [inline pkgs/build-support/mkshell/default.nix:48]: WDYT of something like:

```suggestion
    echo " WARNING: the existence of this path is not guaranteed." >>$out
```

Comment (zimbatm) [reply pkgs/build-support/mkshell/default.nix]: The warning is already scary enough with your changes so we don't have to add even more text.

**Comment by zimbatm** [comment]
Waiting for a last round of reviews.

**Review by roberth — APPROVED** [review]

**Comment by rvl** [comment]
Very helpful, thanks @zimbatm and @roberth - I can remove my `nobuildPhase` overrides now.

**Comment by Atry** [comment]
`buildEnv` is not used any where. Shall we delete it?

**Comment by charles-dyfis-net** [comment]
There's still equivalent logic to the restriction this pulls elsewhere in nixpkgs:

```console
pkgs/build-support/build-fhsenv-bubblewrap/default.nix:      echo >&2 "*** User chroot 'env' attributes are intended for interactive nix-shell sessions, not for building! ***"
pkgs/build-support/build-fhsenv-chroot/default.nix:      echo >&2 "*** User chroot 'env' attributes are intended for interactive nix-shell sessions, not for building! ***"
pkgs/development/interpreters/lua-5/wrapper.nix:          echo >&2 "*** lua 'env' attributes are intended for interactive nix-shell sessions, not for building! ***"
pkgs/development/interpreters/octave/build-env.nix:        echo >&2 "*** octave 'env' attributes are intended for interactive nix-shell sessions, not for building! ***"
pkgs/development/interpreters/python/wrapper.nix:          echo >&2 "*** Python 'env' attributes are intended for interactive nix-shell sessions, not for building! ***"
pkgs/development/ruby-modules/bundled-common/default.nix:            echo >&2 "*** Ruby 'env' attributes are intended for interactive nix-shell sessions, not for building! ***"
```

**Pull request merged**